File: CsharpProjects/myLeetCode/Backtracking/77. Combinations.cs
// 77. Combinations
// https://leetcode.com/problems/combinations/

#include <vector>
#include "../printUtils.hpp"

class Solution {
public:
    std::vector<std::vector<int>> combine(int n, int k) {
        std::vector<std::vector<int>> result;
        std::vector<int> curr;
        backtrack(result, curr, n, k, 1);
        return result;
    }

private:
    void backtrack(std::vector<std::vector<int>>& result, std::vector<int>& curr, int n, int k, int start) {
        if (curr.size() == k) {
            result.push_back(curr);
            return;
        }
        
        // Pruning condition: i <= n - (k - curr.size()) + 1
        for (int i = start; i <= n - (k - curr.size()) + 1; i++) {
            curr.push_back(i);
            backtrack(result, curr, n, k, i + 1);
            curr.pop_back();
        }
    }
};

int main() {
    printSet(Solution().combine(4, 2));

    return 0;
}

File: C++/代码随想录/Backtracking/77. Combinations.py
def combine(n, k):
    def backtrack(start, path):
        if len(path) == k:
            result.append(path[:])
            return
        for i in range(start, n + 1):
            path.append(i)
            backtrack(i + 1, path)
            path.pop()

    result = []
    backtrack(1, [])
    return result

# 测试用例
n = 4
k = 2
print(combine(n, k))
